src/polymer/root-selector-to-html.ts

```diff
@@ -0,0 +1,106 @@
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as dom5 from 'dom5';
+import {Document, ParsedCssDocument, Severity, Warning} from 'polymer-analyzer';
+import * as shady from 'shady-css-parser';
+
+import {registry} from '../registry';
+import {Rule} from '../rule';
+import {getDocumentContaining, stripIndentation, stripWhitespace} from '../util';
+
+const p = dom5.predicates;
+const isCustomStyle = p.AND(
+    p.hasTagName('style'),
+    p.OR(
+        p.hasAttrValue('is', 'custom-style'),
+        (node: dom5.Node) => !!(
+            node.parentNode && p.hasTagName('custom-style')(node.parentNode))));
+const isElementStyle = p.AND(
+    p.hasTagName('style'),
+    (node: dom5.Node) => !!(
+        node.parentNode && node.parentNode.nodeName === '#document-fragment'));
+
+class RootSelectorToHtml extends Rule {
+  code = 'root-selector-to-html';
+  description = stripIndentation(`
+      Warns when using :root inside an element's template or custom-style
+  `);
+
+  async check(document: Document) {
+    return [
+      ...this.generateWarnings(
+          document, isCustomStyle, 'html'),  // Check custom styles
+      ...this.generateWarnings(
+          document, isElementStyle, ':host > *'),  // Check element styles
+    ];
+  }
+
+  private generateWarnings(
+      document: Document, predicate: dom5.Predicate, replacementText: string) {
+    const warnings: Warning[] = [];
+
+    const styleTags = dom5.queryAll(
+        document.parsedDocument.ast,
+        predicate,
+        [],
+        dom5.childNodesIncludeTemplate);
+    if (styleTags.length === 0) {
+      return warnings;
+    }
+
+    for (const style of styleTags) {
+      const sourceRange =
+          document.parsedDocument.sourceRangeForNode(style.childNodes[0]);
+      if (sourceRange === undefined) {
+        continue;
+      }
+      const containingDoc =
+          getDocumentContaining(sourceRange, document) as ParsedCssDocument;
+      if (containingDoc === undefined) {
+        continue;
+      }
+
+      for (const node of containingDoc) {
+        if (node.type !== shady.nodeType.ruleset) {
+          continue;
+        }
+
+        const deprecatedRegex = /:root/;
+        const match = node.selector.match(deprecatedRegex);
+        if (match === null) {
+          continue;
+        }
+
+        const start = node.range.start + match.index!;
+        const sourceRange = containingDoc.sourceRangeForShadyRange(
+            {start, end: start + match[0].length});
+
+        warnings.push(new Warning({
+          parsedDocument: document.parsedDocument,
+          code: this.code,
+          severity: Severity.WARNING, sourceRange,
+          message: stripWhitespace(`
+            Root should no longer be used
+          `),
+          fix: [{range: sourceRange, replacementText}]
```

Comment:
We need to replace just the `:root` part, so I think we'll want to get the replacementText via a regex replace of `:root` with the passed in replacementText

Reply:
I'm not sure what you mean on this

Reply:
How does this handle a selector like `:root foo`?

I think it would turn it into either `html` or `:host > *` but (I think) we want it to emit `html foo` or `:host > * foo`